dev/devicelab/bin/tasks/flutter_engine_group_performance.dart

```diff
@@ -0,0 +1,114 @@
+}
+
+List<String> _addComponent(List<String> components, String value) {
+  final List<String> result = <String>[];
+  result.addAll(components);
+  result.add(value);
+  return result;
+}
+
+List<String> _addComponents(List<String> components, List<String> values) {
+  final List<String> result = <String>[];
+  result.addAll(components);
+  result.addAll(values);
+  return result;
+}
+
+Future<void> _run(String command, List<String> args, String cwd) async {
+  print('$command: $args');
+  final Process process =
+      await Process.start(command, args, workingDirectory: cwd);
+  final Future<dynamic> stdoutStream = stdout.addStream(process.stdout);
+  final Future<dynamic> stderrStream = stderr.addStream(process.stderr);
+  final int exitCode = await process.exitCode;
+  if (exitCode != 0) {
+    throw Exception('command "$command $args" had exit code: $exitCode');
+  }
+  await stdoutStream;
+  await stderrStream;
+}
+
+Future<void> main() async {
+  try {
+    final List<String> scriptPathComponents =
+        _splitAbsPath(Platform.script.path);
+    final List<String> multipleFluttersPath =
+        scriptPathComponents.sublist(0, scriptPathComponents.length - 4);
+    multipleFluttersPath.addAll(<String>['benchmarks', 'multiple_flutters']);
+    final String modulePath =
+        _joinAbsPath(_addComponent(multipleFluttersPath, 'module'));
+    final String androidPath =
+        _joinAbsPath(_addComponent(multipleFluttersPath, 'android'));
+
+    await _run('flutter', <String>['pub', 'get'], modulePath);
+    await _run('./gradlew', <String>['assembleRelease'], androidPath);
+
+    final String apkPath = _joinAbsPath(_addComponents(
+        multipleFluttersPath, <String>[
+      'android',
+      'app',
+      'build',
+      'outputs',
+      'apk',
+      'release',
+      'app-release.apk'
+    ]));
+    await _run(
+        'adb', <String>['install', '-r', apkPath], Directory.current.path);
+    await _run(
+        'adb',
+        <String>[
+          'shell',
+          'am',
+          'start',
+          '-n',
+          '$_bundleName/$_bundleName.$_activityName'
+        ],
+        Directory.current.path);
+    await Future<void>.delayed(const Duration(seconds: 10));
+    final ProcessResult meminfoResult = await Process.run(
+        'adb', <String>['shell', 'dumpsys', 'meminfo', _bundleName]);
+    if (meminfoResult.exitCode != 0) {
+      throw Exception('meminfo returned exit code: ${meminfoResult.exitCode}');
+    }
+    assert(meminfoResult.exitCode == 0);
+    final RegExp regex = RegExp(r'TOTAL:\s+(\d+)');
+    final int totalMemory =
+        int.parse(regex.firstMatch(meminfoResult.stdout.toString()).group(1));
+    print('''
+{
+  "success": true,
+  "data": {
+    "totalMemory": $totalMemory
+  },
+  "detailFiles": null,
+  "benchmarkScoreKeys": ["totalMemory"],
+}
+''');
+  } catch (e) {
+    print('''
+{
+  "success": false,
+  "reason": "${e.toString()}"
+}
+''');
```

Comment:
@keyonghan This looks correct, right?  Where should I write the file?  It wasn't apparent to me.

Reply:
From the test side, what you need to do is return a [`TaskResult`](https://devops-console-oss.corp.google.com/flutter/flutter/+/master:dev/devicelab/lib/framework/task_result.dart;l=9): where the `data` would be the metrics you want to collect.

For example:
The test `analyzer_benchemark`, is called via [`analyzer_benchmark.dart`](https://devops-console-oss.corp.google.com/flutter/flutter/+/master:dev/devicelab/bin/tasks/analyzer_benchmark.dart), then is forwarded to call [`analyzerBenchmarkTask`](https://devops-console-oss.corp.google.com/flutter/flutter/+/master:dev/devicelab/lib/tasks/analysis.dart;l=23) function, which returns the metrics as `TaskResult`.

So the workflow is like below:

- A task is called/executed via the [devicelab recipe ](https://devops-console-oss.corp.google.com/flutter/recipes/+/master:recipes/devicelab/devicelab_drone.py;l=88)
- The [`bin/run`](https://devops-console-oss.corp.google.com/flutter/flutter/+/master:dev/devicelab/bin/run.dart) will be called. If argument [`--result-file`](https://devops-console-oss.corp.google.com/flutter/recipes/+/master:recipes/devicelab/devicelab_drone.py;l=51) is specified, the test will write `TaskResult` to this local file.
- If [`upload_metrics`](https://devops-console-oss.corp.google.com/flutter/recipes/+/master:recipes/devicelab/devicelab_drone.py;l=106) is further specified, the local file contents will be uploaded to cocoon/Skia.

From test side, you only need to have the test return a `TaskResult`. Writing to local file, and upload to cocoon/skia is managed in [recipe](https://devops-console-oss.corp.google.com/flutter/recipes/+/master:recipes/devicelab/devicelab_drone.py) and [builder config side](https://devops-console-oss.corp.google.com/flutter/infra/+/master:config/devicelab_config.star).

Reply:
Okay, thanks.  I think I've got it.